deku/protocol/address.re

```diff
@@ -1,23 +1,55 @@
 open Crypto;
-open Key_hash;
 
-[@deriving ord]
-type t = Key_hash.t;
+module Implicit = {
+  include Key_hash;
 
-let of_wallet = of_key;
-let pubkey_matches_wallet = (key, t) => equal(of_key(key), t);
+  let of_wallet = of_key;
+  let pubkey_matches_wallet = (key, t) => equal(of_key(key), t);
 
-let make = () => {
-  let (key, pub_) = Ed25519.generate();
-  let wallet_address = of_wallet(Ed25519(pub_));
+  let make = () => {
+    let (key, pub_) = Ed25519.generate();
+    let wallet_address = of_wallet(Ed25519(pub_));
 
-  (Secret.Ed25519(key), wallet_address);
+    (Secret.Ed25519(key), wallet_address);
+  };
+  let to_key_hash = t => t;
+  let of_key_hash = t => t;
 };
-let to_key_hash = t => t;
-let of_key_hash = t => t;
 
-let to_string = to_string;
-let of_string = of_string;
+[@deriving (yojson, ord)]
+type t =
+  | Implicit(Implicit.t)
+  | Originated(Operation_hash.t);
 
-let to_yojson = to_yojson;
-let of_yojson = of_yojson;
+let of_implicit = implicit => Implicit(implicit);
+let to_implicit = t =>
+  switch (t) {
+  | Implicit(implicit) => Some(implicit)
+  | _ => None
+  };
+
+let to_string = t =>
+  switch (t) {
+  | Implicit(t) => "dz1" ++ Key_hash.to_string(t)
```

Comment:
Key_hash will have tz1 when stringified will look like `dz1tz1<rest>` is this intended? Should we really support tz2 and tz3 inside deku as internal addresses?